alacritty/src/renderer/box_drawing.rs

```diff
@@ -0,0 +1,584 @@
+struct RGBPixel {
+    _r: u8,
+    _g: u8,
+    _b: u8,
+}
+
+/// Canvas which is used for simple line drawing operations.
+struct Canvas {
+    /// Canvas width.
+    width: usize,
+
+    /// Canvas height.
+    height: usize,
+
+    /// Canvas buffer we draw on.
+    buffer: Vec<RGBPixel>,
+}
+
+impl Canvas {
+    /// Builds new `Canvas` for line drawing with the given `width` and `height` with default color.
+    fn new(width: usize, height: usize) -> Self {
+        let buffer = vec![RGBPixel::default(); width * height];
+        Self { width, height, buffer }
+    }
+
+    /// Vertical center of the `Canvas`.
+    fn center_v(&self) -> f32 {
+        self.height as f32 / 2.
+    }
+
+    /// Horizontal center of the `Canvas`.
+    fn center_h(&self) -> f32 {
+        self.width as f32 / 2.
+    }
+
+    /// Canvas underlying buffer for direct manipulation
+    fn buffer_mut(&mut self) -> &mut [RGBPixel] {
+        &mut self.buffer
+    }
+
+    /// Gives bounds for horizontal straight line on `y` with `stroke_size`.
+    fn h_line_bounds(&self, y: f32, stroke_size: usize) -> (f32, f32) {
+        let start_y = max((y - stroke_size as f32 / 2.) as i32, 0) as f32;
+        let end_y = min((y + stroke_size as f32 / 2.) as i32, self.height as i32) as f32;
+
+        (start_y, end_y)
+    }
+
+    /// Gives bounds for vertical straight line on `y` with `stroke_size`.
+    fn v_line_bounds(&self, x: f32, stroke_size: usize) -> (f32, f32) {
+        let start_x = max((x - stroke_size as f32 / 2.) as i32, 0) as f32;
+        let end_x = min((x + stroke_size as f32 / 2.) as i32, self.width as i32) as f32;
+
+        (start_x, end_x)
+    }
+
+    /// Draws a horizontal straight line from (`x`, `y`) of `size` with the given `stroke_size`.
+    fn draw_h_line(&mut self, x: f32, y: f32, size: f32, stroke_size: usize) {
+        let (start_y, end_y) = self.h_line_bounds(y, stroke_size);
+        self.draw_rect(x, start_y as f32, size, (end_y - start_y) as f32, COLOR_FILL);
+    }
+
+    /// Draws a vertical straight line from (`x`, `y`) of `size` with the given `stroke_size`.
+    fn draw_v_line(&mut self, x: f32, y: f32, size: f32, stroke_size: usize) {
+        let (start_x, end_x) = self.v_line_bounds(x, stroke_size);
+        self.draw_rect(start_x as f32, y, (end_x - start_x) as f32, size, COLOR_FILL);
+    }
+
+    /// Draws a rect from the (`x`, `y`) of the given `width` and `height` using `color`.
+    fn draw_rect(&mut self, x: f32, y: f32, width: f32, height: f32, color: RGBPixel) {
+        let start_x = x as usize;
+        let end_x = min((x + width) as usize, self.width);
+        let start_y = y as usize;
+        let end_y = min((y + height) as usize, self.height);
+        for y in start_y..end_y {
+            let y = y * self.width;
+            self.buffer[start_x + y..end_x + y].fill(color);
+        }
+    }
+
+    /// Naive arbitrary line drawing from (`from_x`, `from_y`) to (`to_x`, `to_y`).
```

Comment:
It's not really arbitrary, is it? Kinda makes it sound like you're going in a zigzag or whatever.